### DISCUSSION TITLE
Do AppState and AppAction need to be Equatable?

### DISCUSSION THREAD

**Discussion by gohanlon** [pointfreeco/isowords#110] [General]
I noticed that Isowords still builds if the conformance to `Equatable` is removed from `AppAction`:

https://github.com/pointfreeco/isowords/blob/d9d4e7742e4230d7a92bde72853e4cd76528480f/Sources/AppFeature/AppView.swift#L67

Then I noticed that the same is true for `AppState`:

https://github.com/pointfreeco/isowords/blob/d9d4e7742e4230d7a92bde72853e4cd76528480f/Sources/AppFeature/AppView.swift#L18

Since they're not technically needed, would removing these `Equatable` conformances make sense? Or, perhaps including them is a good practice?

Either way, this is clearly a very minor issue/nit. I just happened to notice this while studying the (awesome!) codebase and am curious to hear others' thoughts. Thank you!

**Comment by filblue** [comment]
Hi @gohanlon, the [state](https://github.com/pointfreeco/swift-composable-architecture/blob/f586c105ed5652dcd112c0e746d1d82e3484791d/Sources/ComposableArchitecture/TestSupport/TestStore.swift#L306) and [action](https://github.com/pointfreeco/swift-composable-architecture/blob/f586c105ed5652dcd112c0e746d1d82e3484791d/Sources/ComposableArchitecture/TestSupport/TestStore.swift#L161) are required to be `Equatable` in order to be testable via `TestStore`. If you remove the conformance, the tests would not build.

**Comment by mbrandonw** [comment]
Yep, as @filblue said, the `Equatable` conformance are necessary for testing. The Composable Architecture comes with test helpers that allow you to test the full end-to-end lifecycle of sending user actions, executing effects and receiving data from effects. In order for that helper to work both state and actions need to be equatable. If you want to learn more about this helper you can watch our [videos](https://www.pointfree.co/collections/composable-architecture/testing) where we develop the helper from scratch.

I suppose that the `Equatable` conformances could be moved to the test modules since they are not needed in the application module, but I believe then the conformances will not be automatically synthesized by Swift, so we'd have to fully write it out ourselves.

**Comment by gohanlon** [comment]
Ah, of course! Thank you @filblue and @mbrandonw.

I was pondering how to go about creating a TCA module that would wrap an Apple API that's SwiftUI view-based and provides arguments that are not `Equatable`. I managed to consider that testing and mocking while using Apple's types for action case-associated data would be problematic even if they were `Equatable`—Apple's types seem hard or impossible to instantiate—but the `Equatable` conformance requirement on actions being due to `TestStore` somehow didn't occur to me. 🤦🏻‍♂️

I also remember a PointFree episode on containing such un-construct-able Apple Framework types—part of the Designing Dependencies series, I think. I'll go re-watch those. Thanks again!

**Reply by mbrandonw** [reply]
> I also remember a PointFree episode on containing such un-construct-able Apple Framework types—part of the Designing Dependencies series, I think. I'll go re-watch those. Thanks again!

Yeah exactly, it happens in part 2 of that series, starting around [16:30](https://www.pointfree.co/episodes/ep112-designing-dependencies-reachability#t988).
